server/lib/pbench/process_tb.py

```diff
@@ -65,8 +65,15 @@ def _results_push(controller: str, tb: Path, token: str):
         """
 
         res = subprocess.run(
-            ["pbench-results-push", controller, tb, token], capture_output=True
+            ["pbench-results-push", controller, tb, f"--token={token}"],
+            capture_output=True,
+            env={
+                "PATH": "/opt/pbench-agent/util-scripts",
+                "PYTHONPATH": "/opt/pbench-agent/lib/python3.9/site-packages:/opt/pbench-agent/lib64/python3.9/site-packages:/opt/pbench-agent/lib",
+                "_PBENCH_AGENT_CONFIG": "/opt/pbench-agent/config/pbench-agent.cfg",
+            },
         )
```

Comment:
That value of `PYTHONPATH` is highly suspect.  While I have no doubt that it works for you today on the system where you are testing, embedding a dependency on Python3.9 this deeply into our code seems like a really bad idea, even if it were the correct value in general.

I believe that `pbench-results-push` is an Agent tool, and I believe that the stated requirement for the Agent is Python 3._**6**_, so we almost certainly should not be (explicitly) referencing `python3.9` here.

Instead, either we should be determining what version of Python the `pbench-results-push` is going to use or we should be looking up which path matches `/opt/pbench-agent/lib/python3*/site-packages` and use the result accordingly.

Reply:
I've looked at this code several times and debated what to say.

1. On one hand, the put-shim server is a unique end-of-life instance of 0.69;
2. On the other hand, some level of generality isn't a bad thing.

The `PATH` and `PYTHONPATH` normally from from `/etc/profile.d/pbench-agent.sh`, and one possibility would be to fork a bash subprocess which sources that file before using the agent command.

Another would be to pull pieces of that to properly configure the settings, probably by explicitly pulling the config file with a `ConfigParser` instance and reading the top directory and other relevant parameters, along with a search for the `site-packages` we actually have.

The question is wow much of that effort is really justified here... in terms of our one-off put-shim, maybe none of it. *But*, at the absolute best case we're likely to be deploying this multiple times before we get everything right, and having generalized code in the repo makes sense.

As for the 3.6 vs 3.9 ... I wondered about that, too. What did we choose as the base Python for the 0.69 server? Did we bump that to 3.9 as well? Somehow I thought that was an earlier version. And, on this hybrid put-shim configuration, do we want the agent code to be running on the same version of Python, or on a separate version?

Reply:
One other thing: I think the environment of `pbench-results-push` needs that `PATH` to find the tar-building utilities etc, but `put-shim` also needs to be able to find `pbench-results-push` itself, so it might not be enough to set the `PATH` only in the subprocess's environment: we need to set it in the environment of the caller. Alternatively, it can call `pbench-results-push`  using an absolute path.

Reply:
> If we are OK with the (barely adequate) method that's used in `/etc/profile.d/pbench-agent.sh` to set `PYTHONPATH` then we can do the same thing here

Yeah, that's pretty much what I had in mind, although I do not like the fact that that code appears to be ready to put site package directories from _more than one_ Python release into the `PYTHONPATH`...that seems like a potential disaster....

However, the site package directories appear to be specific to the Agent, and presumably they are set up when the Agent is installed, so hopefully there is only one Python-version's site package directory in play here (and, the Server installation shouldn't impact this).  If we assume that that is the case, then we can have the code search for it and use it...and if it finds more than one, then it can issue an error and we'll deal with the situation if it arises.

Reply:
> As for the 3.6 vs 3.9 ... I wondered about that, too. What did we choose as the base Python for the 0.69 server? Did we bump that to 3.9 as well?

the python version that is considered in put-shim server rpm is 3.9. [pbench-server spec file](https://download.copr.fedorainfracloud.org/results/portante/pbench-0.69/epel-8-x86_64/05036498-pbench-server/pbench-server.spec)

Reply:
Right, but the Pbench Agent is a separate installation, and there is nothing which says that we cannot have more than one version of Python on the same system.  So, the challenge to this approach is that the Pbench Server, which is running under Python 3.9, has to be able to predict which version of Python the Pbench Agent will be run under (or, at least, it has to pick the right site packages for its invocation...hopefully there is only one to choose from).

A better approach would be to get the Server's invocation of the Agent to use the defaults which the Agent installation set up, so that the Server won't have to guess or choose anything about the configuration.